File: CS/IntsValidator.cs
/*
 * Team 3
 * AJ Downey, Joshua Barbee, Wei Wei Chien
 */

namespace Defend483CS
{
    internal class IntsValidator
    {

        private const string _title =
            "2 integers:";

        private const string _firstIntTitle =
            "First integer:";
        private const string _secondIntTitle =
            "Second integer:";
        private static readonly string _prompt =
            "Enter one integer in the range [" + int.MinValue + ", " +
            int.MaxValue + "]." + Environment.NewLine + "The integer can " +
            "include a sign (+ or -) and cannot include commas:";

        internal static (int, int, long, long) ReadAndAddAndMultiply()
        {
            (int first, int second) = InputManager.ReadTwo<int>(
                _title,
                _firstIntTitle, _secondIntTitle,
                ReadInt
            );

            return (
                first, second,
                Add(first, second), Multiply(first, second)
            );
        }

        // Param not in-mode ref for compiler to allow passing
        // as Func<string, int> param
        private static int ReadInt(string theSubtitle)
        {
            int result = 0; // Guaranteed to be reassigned
            InputManager.ReadStringUntilValid(
                _prompt,
                theTitle: theSubtitle,
                theValidationFunction: TryParseWrapper
            );

            return result;

            // Can't use in param since used as method group
            bool TryParseWrapper(string theString) =>
                int.TryParse(theString, out result);
        }

        private static long Add(in int theFirstInt, in int theSecondInt)
        {
            return ((long)theFirstInt) + theSecondInt;
        }

        private static long Multiply(in int theFirstInt, in int theSecondInt)
        {
            return ((long)theFirstInt) * theSecondInt;
        }
    }
}

File: Java/defend/IntsValidator.java
/*
 * Team 3
 * AJ Downey, Joshua Barbee, Wei Wei Chien
 */

package defend;

public class IntsValidator {

    private static final String INTS_TITLE =
            "2 integers:";

    private static final String FIRST_INT =
            "First integer:";
    private static final String SECOND_INT =
            "Second integer:";
    private static final String INTS_PROMPT =
            "Enter one integer in the range [" + Integer.MIN_VALUE + ", " +
            Integer.MAX_VALUE + "].\nThe integer can include a sign (+ or " +
            "-) and cannot include commas:";

    static int[] read() {
        final int[] ints = new int[2];

        System.out.println(INTS_TITLE);
        ints[0] = readInt(FIRST_INT);
        ints[1] = readInt(SECOND_INT);

        return ints;
    }

    static long add(final int[] theInts) {
        return ((long) theInts[0]) + theInts[1];
    }

    static long multiply(final int[] theInts) {
        return ((long) theInts[0]) * theInts[1];
    }

    private static int readInt(final String thePrePrompt) {
        while (true) {
            System.out.println(thePrePrompt);
            System.out.println(INTS_PROMPT);

            try {
                return Integer.parseInt(
                        InputManager.readLineAndPrintBlankLine()
                );
            } catch (NumberFormatException e) {
                InputManager.printInputInvalid();
            }
        }
    }
}
